File: scala/19.scala
object Solution {
    def removeNthFromEnd(head: ListNode, n: Int): ListNode = {
        val dummyHead = ListNode(0, head)
        var ptr = dummyHead
        for (i <- 0 until n+1) ptr = ptr.next
        var nPtr = dummyHead
        while (ptr != null) {
            nPtr = nPtr.next
            ptr = ptr.next
        }
        nPtr.next = nPtr.next.next
        dummyHead.next
    }
}

File: go/19.go
func removeNthFromEnd(head *ListNode, n int) *ListNode {
	dummyHead := &ListNode{Next: head}
	remove(dummyHead, n)
	return dummyHead.Next
}

func remove(head *ListNode, n int) int {
	if head == nil {
		return 0
	}
	length := 1 + remove(head.Next, n)
	if length == n+1 {
		head.Next = head.Next.Next
	}
	return length
}